tests/test_locks.py

```diff
@@ -0,0 +1,65 @@
+import gc
+
+import pytest
+
+from loguru import logger
+
+
+class CyclicReference:
+    """A minimal cyclic reference.
+
+    Cyclical references are garbage collected using the generational collector rather than
+    via reference counting. This is important here, because the generational collector runs
+    periodically, meaning that it is hard to predict when the stack will be overtaken by a
+    garbage collection process - but it will almost always be when allocating memory of some
+    kind.
+
+    When this object is garbage-collected, a log will be emitted.
+    """
+
+    def __init__(self, _other: "CyclicReference" = None):
+        self.other = _other or CyclicReference(_other=self)
+
+    def __del__(self):
+        logger.info("tearing down")
+
+
+@pytest.fixture()
+def _remove_cyclic_references():
+    """Prevent cyclic isolate finalizers bleeding into other tests."""
+    try:
+        yield
+    finally:
+        for generation in range(3):
+            gc.collect(generation=generation)
+
+
+def test_no_deadlock_on_generational_garbage_collection(_remove_cyclic_references):
+    """Regression test for https://github.com/Delgan/loguru/issues/712
+
+    Assert that deadlocks do not occur when a cyclic isolate containing log output in
+    finalizers is collected by generational GC, during the output of another log message.
+    """
+
+    # GIVEN a sink which assigns some memory
+    output = []
+
+    def sink(message):
+        # This simulates assigning some memory as part of log handling. Eventually this will
+        # trigger the generational garbage collector to take over here.
+        # You could replace it with `json.dumps(message.record)` to get the same effect.
+        _ = [dict() for _ in range(20)]
```

Comment:
I actually don't understand why do we need to trigger the garbage collector this way instead of calling `gc.collect()`? Also, do we actually need to the cyclic references? Isn't the problem related to `__del__` alone, meaning calling `del some_object` triggers the deadlock by itself as long as `some_object.__del__` tries to log a message (and that the GC runs instantly)?

Reply:
Functionally that works!

I based this on the example on the issue, and I avoided explicitly calling `gc.collect` there to demonstrate how this could be triggered reliably with seemingly innocent operations here.

In short, I think this demonstrates how easy it would be to trigger GC at this point, for readers who are not familiar with how the generational GC works.

Happy to switch to `gc.collect` if you'd prefer?

Reply:
Given the unpredictable nature of the garbage collector, I would indeed prefer to call `gc.collect()` to guarantee the test behaves as expected regardless of `gc` configuration and possible change of the `gc` implementation in the future.

It's a good idea to give a concrete example, though. I would suggest to leave it as a comment and/or link to the related ticket you created.